Saving books in a local file cache + util functions

Jatewo opened the pull request:
This pull request introduces a local file caching system for downloaded book texts, refactors file-saving logic for ciphers, and adds comprehensive tests for the new file utility functions. The changes improve efficiency by preventing redundant downloads, standardize file output locations, and enhance test coverage for file operations.

**Caching and file utility enhancements:**

* Added new utility functions in `utils/files.py` for saving, checking, and retrieving cached books, as well as saving ciphers to JSON files in a standardized directory structure.
* Updated `text_fetching/fetcher.py` to use the new caching utilities: before downloading a book, the code now checks if the book is already cached and loads it from disk if available; otherwise, it saves new downloads to the cache. [[1]](diffhunk://#diff-5a05d3bbb81cf5e7980a07bf7461da70813f41e76847caf6e66975496f0010f8R5) [[2]](diffhunk://#diff-5a05d3bbb81cf5e7980a07bf7461da70813f41e76847caf6e66975496f0010f8R116-R118) [[3]](diffhunk://#diff-5a05d3bbb81cf5e7980a07bf7461da70813f41e76847caf6e66975496f0010f8R142-R143)

**Refactoring and standardization:**

* Refactored `main.py` to use the new `save_cipher` utility for saving ciphers, and standardized cipher output to the `ciphers` directory. [[1]](diffhunk://#diff-2e5ad92c43aa96cc3a9cef6c6aec998b216f1379c43b1f651013d25e55989312R12) [[2]](diffhunk://#diff-2e5ad92c43aa96cc3a9cef6c6aec998b216f1379c43b1f651013d25e55989312L23-R29)
* Updated test code to expect cipher files in the `ciphers` directory and to mock file I/O where appropriate. [[1]](diffhunk://#diff-959a1effd1cb0e25016584deef9be1994af65814f318346bf3a63801b5db1636R18-R40) [[2]](diffhunk://#diff-c0469c5748e9772e6e098a1b279aceb9b03b066fb8f42cf9a818416fdc1bb607R13-R15) [[3]](diffhunk://#diff-c0469c5748e9772e6e098a1b279aceb9b03b066fb8f42cf9a818416fdc1bb607R96-R102)

**Testing improvements:**

* Added a new test suite in `test/utils/test_files.py` to cover all new file utility functions, including edge cases for missing or empty files.
* Added tests for book caching logic in the fetcher, ensuring that cached books are loaded without making HTTP requests.
* Added a test for the `Cipher` class to verify that the difficulty parameter is set correctly.

Jatewo committed:
feat: Add file utilities for caching books and saving ciphers

Introduced src/utils/files.py with functions to cache fetched books and save cipher data. Updated Fetcher to use book caching and main.py to use the new save_cipher utility. Adjusted .gitignore to exclude the books cache directory.

Jatewo committed:
test: Add test for Cipher difficulty parameter

Introduces a test to verify that the Cipher class correctly sets the difficulty attribute when initialized with different difficulty values.

Jatewo committed:
fix: Update test paths for cipher file and add save_book mock

Changed test file paths to use 'ciphers/test_cipher.json' instead of the current directory. Added an autouse fixture to mock 'utils.files.save_book' to prevent actual file operations during tests.

Jatewo committed:
test: Add tests for cached book retrieval in Fetcher

Introduces a new test to verify that Fetcher correctly returns cached book content without making HTTP requests. Also adds an autouse fixture to mock save_book and updates existing tests to mock cache checks, improving test isolation and reliability.

Jatewo committed:
fix: Patch book_is_cached in test_fetch_book_no_format

Added a mock for book_is_cached to return False in test_fetch_book_no_format to ensure the test does not rely on cache state.

Jatewo committed:
test: Add tests for utils.files module

Introduces unit tests for the utils.files module, covering book saving, caching, retrieval, and cipher saving functionalities. Uses pytest fixtures for setup and isolation.

Jatewo committed:
fix: linting

Fantaskink committed:
Remove IDs of unavailable books

Fantaskink reviewed (approved):
LGTM

Mafusn commented on src/utils/files.py:
```diff
@@ -0,0 +1,60 @@
+from encipherment.cipher import Cipher
+
+def save_book(book_id: str, book_text: str) -> None:
```
Isn't it quite crazy to pass an entire book text as a parameter, maybe it should just fetch the book by id

Jatewo replied:
No

Mafusn commented on test/text-fetching/test_fetcher.py:
```diff
@@ -90,6 +93,13 @@ def test_fetch_book_success(mocker):
         "formats": {"text/plain; charset=utf-8": "http://example.com/book.txt"}
     }
 
+    mocker.patch("text_fetching.fetcher.save_book")  # Mock save_book to avoid file I/O
```
Isn't this already mocked in mock_save_book?

Jatewo replied:
Yes, it is 👍

Mafusn commented on test/utils/test_files.py:
```diff
@@ -0,0 +1,59 @@
+	yield # Cleanup: Nothing needed as tmp_path is automatically cleaned up
+
+def test_save_book(tmp_path):
+	from utils.files import save_book
```
Wouldn't it be cleaner to import all these functions in the top instead of in each test function?

Jatewo replied:
Yea, you're probably right

TheMOHD3886 replied:
I agree with mafuzz

Mafusn reviewed (changes requested):
These questions likely don't need any changes.

Fantaskink committed:
fix: specify encoding technique to prevent errors on windows

Jatewo committed:
fix: Refactor test imports and mocks for file utilities

Removed redundant inline imports in test_files.py by importing utility functions at the top of the file. Also removed an unnecessary mock for save_book in test_fetcher.py, streamlining the test setup.

Jatewo committed:
Merge branch 'save-books' of github.com:SW9-Cryptanalysis/ciphertext-generation into save-books
